File: backend/src/main/java/com/keystone/service/WorkOrderService.java

package com.keystone.service;

import java.util.List;

import com.keystone.dto.ChangeStatusRequest;
import com.keystone.dto.CustomerWorkOrderRequest;
import com.keystone.dto.WorkOrderRequest;
import com.keystone.dto.WorkOrderResponse;
import com.keystone.dto.WorkOrderStatusHistoryResponse;
import com.keystone.enums.Priority;
import com.keystone.enums.WorkOrderStatus;

public interface WorkOrderService {

    // =========================================================
    // WORK ORDER
    // =========================================================

    WorkOrderResponse createWorkOrder(
    		String email, WorkOrderRequest request
    );

    WorkOrderResponse getWorkOrderById(
            Long workOrderId
    );

    List<WorkOrderResponse> getAllWorkOrders();

    WorkOrderResponse updateWorkOrder(
            Long workOrderId,
            WorkOrderRequest request
    );

    void deleteWorkOrder(
            Long workOrderId
    );


    // =========================================================
    // CUSTOMER
    // =========================================================

    WorkOrderResponse createCustomerWorkOrder(
            String email,
            CustomerWorkOrderRequest request
    );

    List<WorkOrderResponse> getMyWorkOrders(
            String email
    );

    WorkOrderResponse updateMyWorkOrder(
            String email,
            Long workOrderId,
            CustomerWorkOrderRequest request
    );

    List<WorkOrderResponse> getByCustomer(
            Long customerId
    );


    // =========================================================
    // DISPATCHER
    // =========================================================

    List<WorkOrderResponse> getWorkOrdersByStatus(
            WorkOrderStatus status
    );

    List<WorkOrderResponse> getWorkOrdersByPriority(
            Priority priority
    );

    List<WorkOrderResponse> getByTechnician(
            Long technicianId
    );

    WorkOrderResponse assignTechnician(
            String dispatcherEmail,
            Long workOrderId,
            Long technicianId
    );

    // =========================================================
    // STATUS
    // =========================================================

    WorkOrderResponse changeStatus(
            Long workOrderId,
            ChangeStatusRequest request
    );


    // =========================================================
    // TECHNICIAN
    // =========================================================

    List<WorkOrderResponse> getMyTechnicianWorkOrders(
            String email
    );

    List<WorkOrderResponse> getPendingTechnicianAssignments(
            String email
    );

    WorkOrderResponse acceptWorkOrder(
            String email,
            Long workOrderId
    );

    WorkOrderResponse startWorkOrder(
            String email,
            Long workOrderId
    );

    WorkOrderResponse holdWorkOrder(
            String email,
            Long workOrderId
    );

    WorkOrderResponse resumeWorkOrder(
            String email,
            Long workOrderId
    );

    WorkOrderResponse completeWorkOrder(
            String email,
            Long workOrderId
    );


    // =========================================================
    // STATUS HISTORY
    // =========================================================

    List<WorkOrderStatusHistoryResponse> getStatusHistory(
            Long workOrderId
    );

    WorkOrderResponse cancelWorkOrder(String email, Long workOrderId);
}

File: frontend/src/services/workOrderService.ts
import api from "@/lib/api";
import type {
  CustomerWorkOrder,
  DispatcherWorkOrder,
  WorkOrderPriority,
  WorkOrderStatus,
} from "@/types/workOrder";

export const workOrderService = {
  // ============================================================
  // DISPATCHER
  // GET ALL WORK ORDERS
  // ============================================================

  getAllWorkOrders: async (): Promise<DispatcherWorkOrder[]> => {
    const response = await api.get<DispatcherWorkOrder[]>(
      "/workorders"
    );

    return response.data;
  },

  // ============================================================
  // DISPATCHER
  // GET WORK ORDER BY ID
  // ============================================================

  getWorkOrderById: async (
    id: number
  ): Promise<DispatcherWorkOrder> => {
    const response = await api.get<DispatcherWorkOrder>(
      `/workorders/${id}`
    );

    return response.data;
  },

  // ============================================================
  // DISPATCHER
  // GET WORK ORDERS BY STATUS
  // GET /api/workorders/status/{status}
  // ============================================================

  getWorkOrdersByStatus: async (
    status: WorkOrderStatus
  ): Promise<DispatcherWorkOrder[]> => {
    const response = await api.get<DispatcherWorkOrder[]>(
      `/workorders/status/${status}`
    );

    return response.data;
  },

  // ============================================================
  // DISPATCHER
  // GET WORK ORDERS BY PRIORITY
  // GET /api/workorders/priority/{priority}
  // ============================================================

  getWorkOrdersByPriority: async (
    priority: WorkOrderPriority
  ): Promise<DispatcherWorkOrder[]> => {
    const response = await api.get<DispatcherWorkOrder[]>(
      `/workorders/priority/${priority}`
    );

    return response.data;
  },

  // ============================================================
  // DISPATCHER
  // ASSIGN TECHNICIAN
  // PATCH /api/workorders/{id}/assign?technicianId={id}
  // ============================================================

  assignTechnician: async (
    workOrderId: number,
    technicianId: number
  ): Promise<DispatcherWorkOrder> => {
    const response = await api.patch<DispatcherWorkOrder>(
      `/workorders/${workOrderId}/assign?technicianId=${technicianId}`
    );

    return response.data;
  },

  // ============================================================
  // CUSTOMER
  // GET MY WORK ORDERS
  // ============================================================

  getMyWorkOrders: async (
    email: string
  ): Promise<CustomerWorkOrder[]> => {
    const response = await api.get<CustomerWorkOrder[]>(
      `/workorders/my?email=${encodeURIComponent(email)}`
    );

    return response.data;
  },

  // ============================================================
  // CUSTOMER
  // CREATE WORK ORDER
  // ============================================================

  createMyWorkOrder: async (
    email: string,
    request: {
      title: string;
      description: string;
      serviceType: string;
      priority: "LOW" | "MEDIUM" | "HIGH";
      scheduledDate: string;
      siteId: number;
    }
  ): Promise<CustomerWorkOrder> => {
    const response = await api.post<CustomerWorkOrder>(
      `/workorders/my?email=${encodeURIComponent(email)}`,
      request
    );

    return response.data;
  },

  // ============================================================
  // CUSTOMER
  // UPDATE WORK ORDER
  // ============================================================

  updateMyWorkOrder: async (
    email: string,
    id: number,
    request: {
      title: string;
      description: string;
      serviceType: string;
      priority: "LOW" | "MEDIUM" | "HIGH";
      scheduledDate: string;
      siteId: number;
    }
  ): Promise<CustomerWorkOrder> => {
    const response = await api.put<CustomerWorkOrder>(
      `/workorders/my/${id}?email=${encodeURIComponent(email)}`,
      request
    );

    return response.data;
  },
};